### ISSUE TITLE
Hydrated framework components in MDX in content collections break `astro build`

### ISSUE THREAD

**Issue by delucis** [withastro/astro#6162]
### What version of `astro` are you using?

2.0.6 (also tested 2.0.2)

### Are you using an SSR adapter? If so, which one?

n/a

### What package manager are you using?

npm/pnpm

### What operating system are you using?

macOS / Stackblitz

### Describe the Bug

There seems to be an issue resolving paths when bundling hydration scripts for framework components used in MDX in content collections. Running into this as part of the docs migration — https://github.com/withastro/docs/pull/2546.

Minimal example that triggers an error when rendered via `<Content />` during `astro build` (`astro dev` works fine):

**src/content/posts/post.mdx**

```mdx
---
title: MDX post with hydrated component
---

import Component from '../../components/PreactComponent';

<Component client:load />
```

Example error stack trace from the docs migration branch:

```
Error: Cannot find the built path for /.../src/components/tabs/Tabs
    at Object.resolve (file:///.../node_modules/.pnpm/astro@2.0.6_lo4g36gygrjg64j6jydpsd7m54/node_modules/astro/dist/core/build/generate.js:252:15)
    at generateHydrateScript (file:///.../dist/chunks/astro.615ddc00.mjs:951:48)
    at renderFrameworkComponent (file:///.../dist/chunks/astro.615ddc00.mjs:1667:24)
    at async renderChild (file:///.../dist/chunks/astro.615ddc00.mjs:1029:11)
    at async RenderTemplateResult.[Symbol.asyncIterator] (file:///.../dist/chunks/astro.615ddc00.mjs:716:7)
    at async renderAstroTemplateResult (file:///.../dist/chunks/astro.615ddc00.mjs:724:20)
    at async renderToString (file:///.../dist/chunks/astro.615ddc00.mjs:757:20)
    at async renderJSXVNode (file:///.../dist/chunks/astro.615ddc00.mjs:1241:31)
    at async Promise.all (index 16)
    at async renderJSX (file:///.../dist/chunks/astro.615ddc00.mjs:1206:10)
```

My best hunch so far is this has to do with path resolution. Notably the example above will work if we replace the relative path with an alias, e.g.:

```js
import Component from '~/components/PreactComponent';
```

However it goes deeper than this. In the reproduction the hydration script fails to build even if the Preact component is imported indirectly via an Astro component (which is what we do in docs):

**src/content/posts/post.mdx**

```mdx
import AstroComp from '~/components/AstroComp.astro'
```

**src/components/AstroComp.astro**

```astro
---
import PreactComp from './PreactComponent';
---

<PreactComp client:load />
```

This will also fail in `generateHydrateScript`, not finding the built path for `PreactComponent`.

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/github-tgvzww

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by matthewp** [comment]
Just tested this locally via a unit test and also the stackblitz example and it seems to be working... Am I missing something? I get `I hydrated on the client!` for all of the pages marked with an ❌

<img width="1391" alt="Screen Shot 2023-02-07 at 2 27 44 PM" src="https://user-images.githubusercontent.com/361671/217345930-39f08ac6-b046-4cd6-a660-657395ca5b74.png">

----

Note that we did just release 2.0.7 that has a bunch of content collection bug fixes, although I'm not aware that this would be one and the stackblitz lockfile does say 2.0.6.

**Comment by matthewp** [comment]
Sorry, I missed that this was prod only 🙈. I do have a local test that runs the build, but I'll investigate what is different about it.

**Comment by matthewp** [comment]
@delucis in the real project where this is failing, do you have mixed usage of relative and `~` together? I can see how this would happen with mixed alias + relative usage, but if all was relative I don't think it would. Just want to confirm that I have the scenario right, in order to make sure im on the right path to fix.

**Comment by delucis** [comment]
In the docs repo it is a relative import inside an alias import, yeah.

Example chain:

https://github.com/withastro/docs/blob/2d2de0d06b528482dd34c9efc7245adbdd38909e/src/content/docs/ar/getting-started.mdx?plain=1#L6
```js
import PackageManagerTabs from '~/components/tabs/PackageManagerTabs.astro'
```

https://github.com/withastro/docs/blob/2d2de0d06b528482dd34c9efc7245adbdd38909e/src/components/tabs/PackageManagerTabs.astro#L1-L5
```astro
---
import Tabs from './Tabs';
---

<Tabs client:visible sharedStore="package-managers">
	<!-- ... -->
</Tabs>
```

☝️ `Tabs` here is the Preact component.

**Comment by matthewp** [comment]
@delucis It does fail, but if you remove all of the other files it doesn't fail. The underlying problem is that if you use aliases and relative imports to reference a client component, it fails because it assumes you only reference them one of the two ways. I have a fix in #6168
